compiler/src/dotty/tools/dotc/transform/TailRec.scala

```diff
@@ -446,14 +459,28 @@ class TailRec extends MiniPhase {
 
         case Return(expr, from) =>
           val fromSym = from.symbol
-          val inTailPosition = !fromSym.is(Label) || tailPositionLabeledSyms.contains(fromSym)
+          val inTailPosition = (!fromSym.is(Label) || tailPositionLabeledSyms.contains(fromSym)) // Label returns are only tail if the label is in tail position
+            && (!fromSym.is(Method) || (fromSym eq method)) // Method returns are only tail if we are looking at the original method
```

Comment:
Since `fromSym` can only be a `Label` or a `Method`, consider rewriting as
```suggestion
          val inTailPosition =
            if fromSym.is(Label) then tailPositionLabeledSyms.contains(fromSym)) // Label returns are only tail if the label is in tail position
            else fromSym eq method // Method returns are only tail if we are looking at the original method
```